pyop2/base.py

```diff
@@ -3556,9 +3536,18 @@ def compute(self):
         with self._parloop_event:
             orig_lgmaps = []
             for arg in self.args:
-                if arg._is_mat and arg.lgmaps is not None:
-                    orig_lgmaps.append(arg.data.handle.getLGMap())
-                    arg.data.handle.setLGMap(*arg.lgmaps)
+                if arg._is_mat:
+                    new_state = {INC: Mat.ADD_VALUES,
+                                 WRITE: Mat.INSERT_VALUES}[arg.access]
+                    for m in arg.data:
+                        m.change_assembly_state(new_state)
+                    arg.data.change_assembly_state(new_state)
+                    if arg.lgmaps is not None:
+                        olgmaps = []
+                        for m, lgmaps in zip(arg.data, arg.lgmaps):
+                            olgmaps.append(m.handle.getLGMap())
+                            m.handle.setLGMap(*lgmaps)
+                        orig_lgmaps.append(olgmaps)
```

Comment:
Can you explain a bit about what is happening here? I think I get what 3545--3550 are doing. Basically you're running over the "block matrices" and gathering the indirection map for local-to-global-block insertion. But why do we keep the original lgmaps? Wouldn't it be sufficient to just keep `olgmaps`?

Reply:
So matrices come with local to global maps that map process local indices to global indices (for insertion). To apply boundary conditions we replace these lgmaps with "temporary" lgmaps that map the boundary rows/columns to negative global indices (and those insertions are then discarded).

These new lgmaps are parloop-specific (and so live on the arg, not the matrix). So here what we're doing is swapping out the lgmaps, then assembly happens (insertions etc...).

Reply:
Added a comment.